README.md
# gisaid-cli
[![Docker Repository on Quay](https://quay.io/repository/broadinstitute/gisaid-cli/status "Docker Repository on Quay")](https://quay.io/repository/broadinstitute/gisaid-cli)

This repo builds a Docker container with [GISAID](https://www.gisaid.org/)-provided CLI tools. Right now, this is only the [CLI uploader](https://www.gisaid.org/fileadmin/gisaid/files/pdfs/CLI_Guide.pdf), which is a python script released by GISAID under a GPLv3 open source license.

gisaid_cli3/CHANGELOG.md
**v3.0.0**

Due to breaking change and feature additions, Major version increment to version to 3. Entrypoint changed to `cli3`.

Breaking change:
- removed `--debug` option from `args_parser`

Non-breaking changes:
- switched `--client_id` and `--username` to `getpass.getpass` to hide terminal text in interactive mode
- major update to `README.md` (hence `README.pdf`)

Feature additions
- added `--force` option for overwrite of `*.authtoken` in `cli3 authenticate` subparser
- added `cli3 template` subparser to provide instructions and a template for uploading


**v2.1.6**

- set `type = str` for username and password argparse objects

**v2.1.5**

- dev branch

**v2.1.4**

- removed timeout from call_api

**v2.1.3**

- updated demo locations to prevent `exception` errors being returned from API

**v2.1.2**

- refactored `uploader` to `cli2` throughout
- pushed tag 2.1.2
- updated readme.md, preparing for release via GISAID portal

**v.2.1.1**

- relative imports are now absolute
- restructured package to allow absolute imports
- pushed tag 2.1.1 (message "release 2.1.1")
- using CHANGELOG.md
- added `uploader/data/known_bugs.txt`
- updated FAQ in README.md

**v2.1.0**

- touch CHANGELOG.md
- pushed (first) tag 2.1.0 (message "release 2.1.0")
- breaking changes:
    - changed the help menu and user interface considerably
        - will no longer work now with calls to version 1 help menu layout
    - frameshifts notifications
        - submitters now have option to use `cli2` to set frameshift notification preferences
- created a python package by refactoring version 1 of the CLI script
- added test data
- added install instructions
- re-wrote documentation
- added runtime clock
- defaults now to appending of log and fails
    - prevents need to guess at what has succeeded (records now preserved in appended logs, not overwritten logs or ephemeral stdout/stderr)
    - prints progress of run to stdout
- removed need to revoke authtokens; instead, now advising users to delete manually and re-create
- removed sys.exit from script when a bad record is found, now continues to end of submission (saving log of fails instead of failing)
    - this is to avoid abortion of entire upload when, for instance, record 5001 of 5002 is bad the submitter does not need to repeat entire submission
- used CI/CD via gitlab CI
- invited beta testers via dev repo, feedback has been received and changes implemented

**v2.0.1**

- moved dev to gitlab
- created package





gisaid_cli3/cli3/__main__.py
#!/usr/bin/env python3

"""
    Upload consensus sequences and metadata to GISAID's EpiCoV
    Copyright (C) 2022 Freunde von GISAID e.V.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

DEFAULT_TOKEN = "./gisaid.authtoken"

from datetime import datetime
from pathlib import PurePath
import sys
import json

STARTTIME = datetime.now()
TAB = "\t"


def get_execution_time():
    "Print total runtime at end of run."
    print(f"\nTotal runtime (HRS:MIN:SECS): {str(datetime.now() - STARTTIME)}")

def args_parser():
    """
    Argument parser setup and build.
    """
    import argparse, textwrap
    parser = argparse.ArgumentParser(prog = "cli3",
                                     formatter_class = argparse.ArgumentDefaultsHelpFormatter,
                                     description="Version 3 Command Line Interface (CLI) for uploading sequence and metadata to GISAID.")
    subparser_args1 = argparse.ArgumentParser(add_help=False)
    subparser_args2 = argparse.ArgumentParser(add_help=False)
    subparser_args3 = argparse.ArgumentParser(add_help=False)
    subparser_args4 = argparse.ArgumentParser(add_help=False)
    subparser_args5 = argparse.ArgumentParser(add_help=False)
    subparser_args6 = argparse.ArgumentParser(add_help=False)
    subparser_args7 = argparse.ArgumentParser(add_help=False)
    subparser_args1.add_argument("--database", help="Target GISAID database.",
                                 default = "EpiCoV",
                                 choices = ["EpiCoV", "EpiFlu", "EpiRSV"])
    subparser_args2.add_argument("--token",
                                 help = "Authentication token.",
                                 default = DEFAULT_TOKEN)
    subparser_args3.add_argument("--username",
                                 help = "Your GISAID username. Leave blank on shared computers.",
                                 type = str)
    subparser_args3.add_argument("--password",
                                 help = "Your GISAID password. Leave blank on shared computers.",
                                 type = str)
    subparser_args3.add_argument("--client_id",
                                 help = "Submitter's client-ID. Leave blank on shared computers. Email clisupport@gisaid.org to request client-ID.")
    subparser_args3.add_argument("--force",
                                 help = "Switch on force overwrite of token given at --token",
                                 action = "store_true")
    subparser_args4.add_argument("--log",
                                 help = "All output logged here.",
                                 default = "./logfile.log")
    subparser_args5.add_argument("--template",
                                 help = "Print submission 'template.csv' file per se.",
                                 action = "store_true")
    subparser_args6.add_argument("--metadata",
                                 help = "The csv-formatted metadata file.",
                                 required = True)
    subparser_args6.add_argument("--fasta",
                                 help = "The fasta-formatted nucleotide sequences file.",
                                 required = True)
    subparser_args6.add_argument("--frameshift",
                                 help = "'catch_none': catch none of the frameshifts and release immediately; 'catch_all': catch all frameshifts and require email confirmation; 'catch_novel': catch novel frameshifts and require email confirmation.",
                                 choices = ["catch_all", "catch_novel", "catch_none"],
                                 default = "catch_all",
                                 required = False)
    subparser_args6.add_argument("--failed",
                                 help = "Name of output file to log failed records.",
                                 default = "./failed.out")
    subparser_args7.add_argument("--proxy",
                                 help = "Proxy-configuration for HTTPS-Request in the form: http(s)://username:password@proxy:port.")


    subparser_modules = parser.add_subparsers(
        title="Sub-commands help", help="", metavar="", dest="subparser_name"
    )

    subparser_modules.add_parser(
        "authenticate",
        help="""Write the authentication token.""",
        description="Write the authentication token.",
        parents=[subparser_args1, subparser_args2, subparser_args3, subparser_args7, subparser_args4],
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    subparser_modules.add_parser(
        "upload",
        help="""Upload sequences and metadata.""",
        description="Perform upload of sequences and metadata to GISAID's curation zone.",
        parents=[subparser_args1, subparser_args2, subparser_args6, subparser_args7, subparser_args4],
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    subparser_modules.add_parser("version",
        help = "Show version and exit.")

    subparser_modules.add_parser("template",
        help = "Print template and formatting instructions",
        description="Print to stdout the formatting instructions for the metadata file, or print the 'template.csv' file per se.",
        parents = [subparser_args1, subparser_args5],
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

    return parser


def main():
    """The main routine"""
    parser = args_parser()
    args = parser.parse_args()

    if parser.parse_args().subparser_name is None:
        print("usage: cli3 -h")
    elif args.subparser_name == "version":
        from cli3 import __version__
        print(f"cli3 version:{TAB}{__version__}")
    elif args.subparser_name == "template":
        from cli3.utils.instructions import INSTRUCTIONS
        from cli3.utils.colors import Colors
        if not args.template:
            for key, value in INSTRUCTIONS[args.database].items():
                print("")
                print(Colors.RED+key+':')
                print(Colors.BLUE+value+"\n")
        else:
            from cli3.utils.template import TEMPLATE
            print(TEMPLATE)
    else:
        from cli3.handler import handle
        try:
            exit_code, logfile = handle(args)
        except KeyboardInterrupt:
            print("user interrupt")
            sys.exit(1)
        if args.log:
            with open(args.log, "a") as f:
                f.write(json.dumps(logfile, indent = 4))
        get_execution_time()
        sys.exit(exit_code) #no need to do this as python does this implicitly at this stage

if __name__ == "__main__":

    import doctest
    doctest.testmod()
    main()



gisaid_cli3/cli3/utils/template.py
#!/usr/bin/env python3

TEMPLATE = """submitter,covv_virus_name,covv_type,covv_passage,covv_collection_date,covv_location,covv_add_location,covv_host,covv_add_host_info,covv_sampling_strategy,covv_gender,covv_patient_age,covv_patient_status,covv_specimen,covv_outbreak,covv_last_vaccinated,covv_treatment,covv_seq_technology,covv_assembly_method,covv_coverage,covv_orig_lab,covv_orig_lab_addr,covv_provider_sample_id,covv_subm_lab,covv_subm_lab_addr,covv_subm_sample_id,covv_authors
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
,,betacoronavirus,,,,,,,,,,,,,,,,,,,,,,,,
"""

gisaid_cli3/cli3/utils/sequencers.py
SEQUENCERS = """BRAND,MODEL
BioZ,BioelectronSeq 4000
ClearLabs,Clear Labs Dx
Illumina,HiSeq 1500
Illumina,HiSeq 2500
Illumina,HiSeq 3000
Illumina,HiSeq 4000
Illumina,HiSeq X
Illumina,iSeq 100
Illumina,MiniSeq
Illumina,MiSeq
Illumina,NextSeq 2000
Illumina,NextSeq 500
Illumina,NextSeq 550
Illumina,NextSeq 550 Dx
Illumina,NovaSeq 6000
MGI,DNBSEQ BGISEQ-500
MGI,DNBSEQ G400
MGI,DNBSEQ G400RS
MGI,DNBSEQ G50
MGI,DNBSEQ T7
Oxford Nanopore Technologies,GridION
Oxford Nanopore Technologies,MinION
Oxford Nanopore Technologies,PromethION
PacBio,Sequel I
PacBio,Sequel II
Thermo Fisher,AB 3500
Thermo Fisher,AB 3500 Dx
Thermo Fisher,AB 3730
Thermo Fisher,AB 3730XL
Thermo Fisher,Ion Torrent
Thermo Fisher,Genexus
Thermo Fisher,Ion PGM Dx
Thermo Fisher,Ion Proton
Thermo Fisher,Ion GeneStudio S5
Thermo Fisher,Ion GeneStudio S5 XL"""

gisaid_cli3/cli3/utils/instructions.py
#!/usr/bin/env python3

"""
    Upload consensus sequences and metadata to GISAID's EpiCoV
    Copyright (C) 2022 Freunde von GISAID e.V.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import pandas as pd
from cli3.utils.sequencers import SEQUENCERS
from io import StringIO


SEQUENCERS = pd.read_csv(StringIO(SEQUENCERS), header=None)
INSTRUCTIONS = {
"EpiCoV": {"Overview": """Fields marked by "**" are required. If a value is missing, use the term "unknown". Fields marked by "***" are required in full. Each row in the submission table is a sample record; each column is a variable.""",
"submitter***": "Your GISAID-registered email address",
"covv_virus_name***": """example hCoV-19/Country/SampleID/YYYY:
Let's break it down into the four parts delineated by the forward slash "/" character:
    - "hCoV-19": despite common usage of virus synonyms such as SARS-CoV-2 or nCoV-19, this first part must remain "hCoV-19" verbatim (to ensure backwards compatibility with EpiCoV db).
    - "Country" is full name of country of sample collection (e.g., Australia), including spaces. For backwards compatibility, the exception being to use "USA" for United States of America.
    - "SampleID" is recommended to be of the format, Loc-Lab-Number, where:
        - Loc is location abbreviation (use abbreviated state or province for location, such as "VIC" for Victoria, Australia, or "CA" for California, USA)
        - Lab is lab name abbreviation (e.g., "CDC" for Centres for Disease Control)
        - Number is sample number or lab code (e.g., 02978, or S47y)
    - "YYYY" is four digit year of sample collection. Note, this must be the same as the YYYY provided in the covv_collection_date value, else a "date inplausible" error will occur

    In this example, the covv_virus_name could be:
        - hCoV-19/Australia/VIC-CDC-02978/2022, or
        - hCoV-19/USA/CA-CDC-S47y/2022, respectively.

    NOTE: covv_virus_name field must match exactly the header of the respective sequence in the fasta file. This is the name key used to match up the sequence value in the sequence file.""",
"covv_type***": """For hCoV-19, this will always be "betacoronavirus".""",
"covv_passage**": """"Original" if the sample was sequenced directly from swabs, otherwise add the name of the cell line (e.g., "Vero") used to culture the specimen.""",
"covv_collection_date***": """Needs to be in the format "YYYY-MM-DD". Required in full.""",
"covv_location***": """Format as "Continent / Country / Region / Sub-region".""",
"covv_add_location": """Use this field to add additional location information. To add postcode or zipcode, add for example "other: postcode 3000".""",
"covv_host**": """For clinical samples, this is "Human". Otherwise add the species name of the organism from which the sample was originally sourced.""",
"covv_add_host_info": """Additional host information. E.g., include the EPI_ISL accession of the first sample in a longitudinal series, add patient co-morbidities, etc.""",
"covv_sampling_strategy": """Could be "longitudinal series" or for GISRS laboratories could be:
    - Sentinel surveillance (ILI),
    - Sentinel surveillance (ARI),
    - Sentinel surveillance (SARI),
    - Non-sentinel-surveillance (hospital),
    - Non-sentinel-surveillance (GP),
    - S gene dropout". """,
"covv_gender**": """Synonym for "Biological sex". Should be "Female", "Male", or "Other".""",
"covv_patient_age**": """Age in years of the person from whom the specimen was collected. May take format other than integer years, for example, "0.5" (i.e., 6 months), "5 days", "7 months". If units are not given, unit is assumed years, for example "35–44" or "35–39", "40–44".""",
"covv_patient_status**": """e.g., "Hospitalized", "Released", "Live", "Deceased". """,
"covv_specimen": """e.g., "Nasopharyngeal swab", "Sputum", "Bronchoalveolar lavage", "Mid-turbinate swab", "Alveolar lavage fluid", "Oropharyngeal swab", "Urine", "Stool", "Cloakal swab", "Organ", "Feces", "Other".""",
"covv_outbreak": """e.g., free text describing date, place, and/or family cluster.""",
"covv_last_vaccinated": """Important for vaccine breakthrough cases. Date and type of vaccination.""",
"covv_treatment": """e.g., Include drug name, dosage.""",
"covv_seq_technology**": "Add the sequencer brand and model.\n\n"+SEQUENCERS.to_csv(sep="\t", index=False, header=0),
"covv_assembly_method": """Freetext describing laboratory and bioinformatic pipeline in brief. For example, PCR primer sets, sequencing library kits and protocol, software and version numbers, reference genome accession against which variant sites were called, e.g. refseq EPI_ISL_402124, Illumina 250bp PE, ARTIC v4.1, minimap2 v2.22-r1101, samtools v1.13, iVar v1.3.1""",
"covv_coverage": """Average depth of coverage (i.e., average read depth) over the length of the reference genome. e.g., "3500x".""",
"covv_orig_lab***": """Full name of laboratory from where sample originated.""",
"covv_orig_lab_addr***": """Complete building address of laboratory from where sample originated.""",
"covv_provider_sample_id": """In-house sample-ID given by originating laboratory.""",
"covv_subm_lab***": """Full name of laboratory submitting this record to GISAID.""",
"covv_subm_lab_addr***": """Complete building address of the submitting laboratory.""",
"covv_subm_sample_id": """In-house sample-ID used by submitting laboratory.""",
"covv_authors***": """Complete list of authors who own the record (i.e., typically submitting and originating author lists combined)."""}
}


gisaid_cli3/cli3/handler.py
#!/usr/bin/env python3

"""
    Upload consensus sequences and metadata to GISAID's EpiCoV
    Copyright (C) 2022 Freunde von GISAID e.V.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""


import argparse, getpass, sys, requests, secrets, hashlib, json, time, csv, os
from itertools import islice
from cli3.utils import __st__

# global configuration
proxy = None
GPS_SERVICE_URL_TEST = "https://gpsapi-test.epicov.org/epi3/gps_api"
GPS_SERVICE_URL_LOCAL = "http://localhost:8080/epi3/gps_api"
GPS_SERVICE_URL = "https://gpsapi.epicov.org/epi3/gps_api"
TAB="\t"
FRAMESHIFTS = {"catch_none": "all", # all frameshifts shall pass
               "catch_all": "none", # no frameshifts shall pass
               "catch_novel": "com"} # only common frameshifts shall pass

def get_service_url(client_type):
    """Get service url."""
    if client_type == "test":
        return GPS_SERVICE_URL_TEST
    elif client_type == "local":
        return GPS_SERVICE_URL_LOCAL
    else:
        return GPS_SERVICE_URL


def call_api(params, debug, client_type):
    """ call the GPS-API """
    global proxy
    body = json.dumps(params)
    try:
        r = requests.post(get_service_url(client_type),
                          data = body,
                          proxies = proxy)
        time.sleep(0.1)
        return r.json()
    except requests.exceptions.ProxyError:
        if debug:
            raise
        else:
            return {"rc": "proxy_error"}
    except KeyboardInterrupt:
        raise
    except:
        if debug:
            raise
        else:
            return {"rc": "connection_error"}

def sha512_hexdigest(inp):
    """Create sha512"""
    hasher = hashlib.sha512()
    hasher.update(inp)
    return hasher.hexdigest()

def request_authentication_token(ctx, client_id, user, password, debug, client_type):
    """
    Request authentication token from GISAID GPS-API.
    Used only for the `authenticate` sub-command (since >=v2.0.1)
    """
    password = password.encode("utf8")
    salt = secrets.token_urlsafe(64)
    hashed = salt + "/" + sha512_hexdigest((salt + sha512_hexdigest(password)).encode("ascii"))

    # resp is response
    resp = call_api({"cmd": "state/auth/get_token",
                     "api": {"version": 1},
                     "ctx": ctx,
                     "client_id": client_id,
                     "login": user,
                     "hash": hashed}, debug,
                     client_type)

    return resp["rc"], resp.get("auth_token"), resp.get("valid_until")


def write_authentication_token_file(ctx, filename, client_id, token, valid_until): #deleted unused 'debug' from args
    """
    Write the authentication token.
    """
    # It seems unusual behaviour that 'tokens' variable is available outside scope of
    # try-except block.  This issue is discussed here: https://bugs.python.org/issue40728
    # and is done because python needs to clean the exception variable outside the block
    # to avoid reference cycles.

    tokens = {}
    tokens[ctx] = {'client-id': client_id,
                   'token': token,
                   'expiry': valid_until,
                  }
    with open(filename, "w") as file:
        file.write(json.dumps(tokens))
    return tokens

def load_authentication_token(ctx, filename, debug):
    """Load the authentication token."""
    try:
        with open(filename, "r") as file:
            return json.loads(file.read())[ctx] # get the row out of json *.authtoken file
    except:
        if debug:
            raise
        else:
            return None

def parse_csv(filename):
    """Parse a csv file."""
    df = None
    # with open(filename, "r") as input_handle:
    #     import pandas as pd
    #     df = pd.parse_csv(input_handle)
    #     print(df)
    file = open(filename, "r", encoding = "utf-8-sig")
    return csv.DictReader(file, delimiter = ",", quotechar = "\"")

def parse_fasta(fn):
    """
    >>> import tempfile
    >>> with tempfile.NamedTemporaryFile() as tmp:
    ...    tmp.write(b'''>EPI_ISL_8747624786432768.1
    ...    ACGACCTAACTGAGGA
    ...    ACACCTAACTGAGAGA''')
    ...    tmp.seek(0)
    ...    sequences = {header: sequence for header, sequence in parse_fasta(tmp.name)}
    ...    print(sequences)
    67
    0
    {'EPI_ISL_8747624786432768.1': 'ACGACCTAACTGAGGAACACCTAACTGAGAGA'}

    """


    seq = None
    header = None
    for row in open(fn, "r", encoding = "utf-8-sig"):
        row = row.strip()
        if row:
            if row[0] == ">":
                if seq:
                    yield header, "".join(seq)
                header = row[1:].lstrip()
                seq = []
            else:
                if ">" in row:
                    raise ValueError("Found '>' in sequence-data")
                seq.append(row)
    if seq:
        yield header, "".join(seq)


def split_every(n, iterable):
    """Split iterable into n-sized chunks."""
    iterable = iter(iterable)
    yield from iter(lambda: list(islice(iterable, n)), [])


def handle(args):
    """
    Handle execution of the workflow:
        authenticate
        upload
    """
    if args.subparser_name == "upload" and args.proxy:
        proxy = {"https": args.proxy,
                 "http": args.proxy}
    client_type = 'live'

    logfile = []
    def log(code, *msg):
        with open(args.log, "a") as f:
            f.write(json.dumps({"code": code, "msg": " ".join(msg)}) + os.linesep)
            print(f"{code}: {', '.join(msg)}")

    failed_writer = None #can avoid this setup in later iterations of cli3 software
    failed_file = None
    def output_faulty(submission):
        if failed_writer:
            failed_writer.writerow(submission)

    # execute CLI-commands
    if args.subparser_name == "authenticate":
        from pathlib import PurePath, Path
        if args.force and Path(args.token).exists():
            print(f"Will overwrite '{Path(args.token)}'.")
        elif (args.force and not Path(args.token).exists()) or (not args.force and not Path(args.token).exists()):
            print(f"Will write authentication token to '{Path(args.token)}'")
        else:
            sys.exit(f"'{args.token}' already exists and force overwrite is off, so doing nothing.")
        if not args.username:
            args.username = getpass.getpass("Enter username: ")
        if not args.password:
            args.password = getpass.getpass("Enter password: ")
        if not args.client_id:
            args.client_id = getpass.getpass("Enter client-ID: ")


        rc, auth_token, valid_until = request_authentication_token(ctx = args.database,
                                                                   client_id = args.client_id,
                                                                   user = args.username,
                                                                   password = args.password,
                                                                   debug = True,
                                                                   client_type = client_type)
        if rc == "ok":
            logfile.append(log("\nok", "user authenticated"))
            logfile.append(log("valid until", time.ctime(valid_until)))
        else:
            print("Response from server:", rc)

        # save it or report error
        if auth_token:
            write_authentication_token_file(args.database, args.token, args.client_id, auth_token, time.ctime(valid_until))
            return 0, logfile
        else:
            return 1, logfile

    elif args.subparser_name == "upload":

        # get the auth-token from the file or from the command line
        from pathlib import PurePath, Path
        if not Path(args.token).exists():
            sys.exit(f"{args.token} does not exist.  Create file using 'cli3 authenticate'.")

        auth_token = load_authentication_token(args.database, args.token, True)
        client_id = auth_token['client-id']
        auth_token = auth_token['token']

        if client_id.startswith("TEST-"):
            client_type = "test"
        elif client_id.startswith("LOCAL-"):
            client_type = "local"

        metadata = parse_csv(args.metadata)

        # prepare a CSV-Write with the same header as the input-file for the failed uploads
        if args.failed: # todo: switch this to pandas, *** args.failed is always true
            failed_file = open(args.failed, "a")
            failed_writer = csv.DictWriter(failed_file,
                                           metadata.fieldnames,
                                           delimiter = ",",
                                           quotechar = "\"",
                                           extrasaction = "ignore")
            failed_writer.writeheader()

        # parse FASTA
        try:
            sequences = {header: sequence for header, sequence in parse_fasta(args.fasta)}
        except ValueError:
            log("fasta_format_error", "upload aborted")
            return 1, logfile

        # check for completeness
        submissions = []
        missing_data = False
        for row in metadata: #column fn does nothing to match with sequence filename, metadata can run without it.
            submission = dict(row)
            if submission["covv_virus_name"] not in sequences:
                log("missing_seq", f"{submission['covv_virus_name']}")
                missing_data = True

            # name in metadata is key, name in fasta is value
            if submission["submitter"] != "Submitter": # a fix to skip demo row if present
                submission["covv_sequence"] = sequences.get(submission["covv_virus_name"])
                submission["_subm_confirmed"] = {"frameshift": FRAMESHIFTS[args.frameshift]}
                submission["_st"] = {"st": f"v {__st__}"}
                submissions.append(submission)

        submission_chunks = split_every(500, submissions)
        count = 0
        ok_count = 0
        for submission_chunk in submission_chunks: # to stop overload on curation, blocks of 500.
            # logon = open session
            resp = call_api({"cmd": "state/session/logon", #cmd is logon
                             "api": {"version": 1},
                             "ctx": args.database,
                             "client_id": client_id,
                             "auth_token": auth_token},
                            debug = True,
                            client_type = client_type)

            time.sleep(0.5) # give system time to save session

            if resp["rc"] == "auth_token_invalid":
                log("invalid_token", "invalid auth-token")
                return 1, logfile
            elif resp["rc"] != "ok":
                log("error", "unexpected response from GPS-API", repr(resp))
                return 1, logfile

            sid = resp["sid"]

            # do uploads
            for submission in submission_chunk:
                submitter = submission["submitter"]
                resp = call_api({"cmd": "data/hcov-19/upload", #cmd is upload
                                 "api": {"version": 1},
                                 "sid": sid,
                                 "ctx": args.database,
                                 "data": submission,
                                 "submitter": submitter,
                                 }, debug = True, client_type = client_type)
                if resp["rc"] == "ok":
                    ok_count += 1
                    log("epi_isl_id", f"{submission.get('covv_virus_name', '')}; {resp['accession_id']}")
                else:
                    output_faulty(submission)
                    if "validation" in resp:
                        log("validation_error", "%s; %s; %s" % (submission.get("covv_virus_name", ""), resp["rc"], json.dumps(resp["validation"])))
                    else:
                        log("upload_error", "%s; %s" % (submission.get("covv_virus_name", ""), resp["rc"])) #resp["rc"]
                count += 1

            # close session
            resp = call_api({"cmd": "state/session/logoff",
                             "api": {"version": 1},
                             "ctx": args.database,
                             "sid": sid},
                            debug = True,
                            client_type = client_type)

        log("upload_count", f"submissions uploaded: {ok_count}")
        log("failed_count", f"submissions failed: {count - ok_count}")

        # close the file for the failed uploads if needed
        if failed_file:
            failed_file.close()


        return 0, logfile
